## Predicting values

Data mining, assignment, 18/04/2021
**Ardit Nela**

We will get to know the practical use of simple methods of supervised modeling or forecasting. The common property of all of these methods is that with the help of random variables (attributes) they model the values of a specific variable,
which we call *class* (in the context of classifying, classifying)
or *response* (in the context of regression). We learnt about the basic differences between contexts in lectures and tutorials.

The practical goals we will pursue are:
* modeling of individual user's responses (responses) with the help of all other users,
* Comparison of supervised modeling methods.

### Data

The description of the MovieLens database remains the same as for the first assignment.

### Preparation of data

For the purposes of this task we will prepare the data as follows:
1. Select $m$ movies with at least 100 views.
2. Select $n$ users who have watched at least 100 movies.
3. Prepare matrix $X$ in the size of $m \times n$, where the lines represent movies and columns represent users. Replace unknown values with $0$.

For each of the selected $n$ users, a regression model will be built, which aims to predict film ratings.

<table>
    <tr style="background-color: white;">
        <td style="border-right: 1px solid #000;"></td>
        <td></td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">$y^{(0)}$</td>
        <td colspan=3 style="text-align:center;">$X^{(0)}$</td>
    </tr>
    <tr style="border-bottom: 1px solid #000;">
        <td style="border-right: 1px solid #000;"></td>
        <td>Film/uporabnik</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">$u_0$</td>
        <td>$u_1$</td>
        <td>$u_2$</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_1}$</td>
        <td>Twelve Monkeys (a.k.a. 12 Monkeys) (1995)</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">0</td>
        <td>0</td>
        <td>2.5</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_2}$</td>
        <td>Dances with  Wolves (1990) </td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">4</td>
        <td>0</td>
        <td>0</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_3}$</td>
        <td>Apollo 13 (1995)</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">0</td>
        <td>2</td>
        <td>0</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_4}$</td>
        <td>Sixth Sense, The (1999)</td><td style="border-right: 1px solid #000; border-left: 1px solid #000;">3</td>
        <td>0</td>
        <td>4</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">$\cdots$</td>
        <td>$\cdots$</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">$\cdots$</td>
        <td>$\cdots$</td>
        <td>$\cdots$</td>
        <td>$\cdots$</td>
    </tr>
</table>

<table>
    <tr style="background-color: white;">
        <td style="border-right: 1px solid #000;"></td>
        <td></td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">$y^{(1)}$</td>
        <td colspan=3 style="text-align:center;">$X^{(1)}$</td>
    </tr>
    <tr style="border-bottom: 1px solid #000;">
        <td style="border-right: 1px solid #000;"></td>
        <td>Film/uporabnik</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">$u_1$</td>
        <td>$u_0$</td>
        <td>$u_2$</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_1}$</td>
        <td>Twelve Monkeys (a.k.a. 12 Monkeys) (1995)</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">0</td>
        <td>0</td>
        <td>2.5</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_2}$</td>
        <td>Dances with  Wolves (1990) </td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">0</td>
        <td>4</td>
        <td>0</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_3}$</td>
        <td>Apollo 13 (1995)</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">2</td>
        <td>0</td>
        <td>0</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">${f_4}$</td>
        <td>Sixth Sense, The (1999)</td><td style="border-right: 1px solid #000; border-left: 1px solid #000;">0</td>
        <td>3</td>
        <td>4</td>
        <td>$\cdots$</td>
    </tr>
    <tr>
        <td style="border-right: 1px solid #000;">$\cdots$</td>
        <td>$\cdots$</td>
        <td style="border-right: 1px solid #000; border-left: 1px solid #000;">$\cdots$</td>
        <td>$\cdots$</td>
        <td>$\cdots$</td>
        <td>$\cdots$</td>
    </tr>
</table>

Data distribution for the model user $u_0$ (top matrix) and the user $u_1$ (bottom matrix).

### Questions

#### 1. Regression (100%)
Set up a regression model for each user. Use one or more methods for learning regression models (linear regression, Ridge, Lasso, etc.).

For each of the $n$ users, select the appropriate column in the data matrix. Therefore, fot the $i$ user we have:

* Response vector $y^{(i)}$,
* Data matrix $X^{(i)}$ containing all columns *except* $i$.
    
For an easier understanding see the above tables. Repeat the test procedure several times (e.g., three times) with the help of the learning and test sets:

* Divide the films that the user viewed *randomly* into 75% (learning set) and 25% (test set).
* Learn the regression model on learning set (select the appropriate rows in $X$ and $y$).
* Evaluate the model on the test set (select the appropriate rows in $X$ and $y$).

Then divide the evaluation score with the number of experiments to get the final score.

Report on the performance of your model. Focus on the following questions:

* Justify an appropriate evaluation score. Does the predict the scores well?
* Rate the models for all $n$ users with the selected evaluation score.

You can split the code needed for each answer into multiple cells.

First we have to filter the top movies and users so we can create a proper matrix.

In [1]:
import pandas as pd
import numpy as np
from csv import DictReader
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split

In [2]:
ratings_raw = pd.read_csv('podatki/ml-latest-small/ratings.csv')
print(ratings_raw.head(20))

#create a dictionary of movieId and title
reader = DictReader(open('podatki/ml-latest-small/movies.csv', 'rt', encoding='utf-8'))
movieNames = dict()
for row in reader:
    movie = int(row["movieId"])
    title = row["title"]
    movieNames[movie] = title

    userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117


In [3]:
movies = set(ratings_raw["movieId"].unique())
m_views = dict()
filteredMovies = dict()
for movie in movies:
    m_views[movie] = len(ratings_raw[ratings_raw["movieId"] == movie])

for movie, ratings in m_views.items():
    if (ratings >= 100):
        filteredMovies[movie] = ratings
print(len(filteredMovies))

151


In [4]:
users = set(ratings_raw["userId"].unique())
u_views = dict()
filteredUsers = dict()
for user in users:
    u_views[user] = len(ratings_raw[ratings_raw["userId"] == user])

for user, ratings in u_views.items():
    if (ratings >= 100):
        filteredUsers[user] = ratings
print(len(filteredUsers))

263


Now we create the matrix.

In [5]:
matrix = []

for movie in filteredMovies:
    matrixRow = []
    tempDF = ratings_raw[ratings_raw["movieId"] == movie]
    for user in filteredUsers:
        if tempDF[tempDF["userId"] == user].empty:
            matrixRow.append(0)
        else:
            matrixRow.append(tempDF[tempDF["userId"] == user]["rating"].item())
    matrix.append(matrixRow)

matrix = np.array(matrix)

This is the function that will filter out the 0s (unrated movies for given user).

In [6]:
def filterEmptyRatings(y, X):
    cond = False
    while not cond:
        for i in range(0, len(y)):
            if y[i] == 0 and i == len(y) - 1:
                y = np.delete(y, i, 0)
                X = np.delete(X, i, 0) 
                cond = True
            elif y[i] == 0:
                y = np.delete(y, i, 0)
                X = np.delete(X, i, 0)
                break
            elif i == len(y) - 1:
                cond = True
    return y, X

This function will model the given matrix. I decided to use Mean Square Error as opposed to r2 or variance because it's the less risky option (data might not be linear).

In [7]:
def createModel(data, model):
    ans = []
    for i in range(np.shape(data)[1]):
        X = np.array(np.zeros(shape=(151, 262)))
        y = data[:, i]
        X = np.delete(data, i, axis=1)
        y, X = filterEmptyRatings(y, X)
        
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
        
        mse = mean_squared_error(y_test, y_pred, squared=False)
        ans.append(mse)
    return ans

Below are all the linear models.

In [8]:
from sklearn.linear_model import LinearRegression
simpleRegression = createModel(matrix, LinearRegression())
print(np.mean([score for score in simpleRegression]))

0.8227808809260485


In [9]:
from sklearn.linear_model import Ridge
ridgeRegression = createModel(matrix, Ridge(0.4))
print(np.mean([score for score in ridgeRegression]))

0.8225102762145889


In [10]:
from sklearn.linear_model import Lasso
lassoRegression = createModel(matrix, Lasso(0.55))
print(np.mean([score for score in lassoRegression]))

0.8398585057336113


In [11]:
from sklearn.linear_model import ElasticNetCV
elasticNet = createModel(matrix, ElasticNetCV(cv=4, max_iter=10000))
print(np.mean([score for score in elasticNet]))

0.8276449874592797


In [12]:
from sklearn.linear_model import LassoCV
lasso = createModel(matrix, LassoCV(tol=0.1, cv=4, max_iter=10000))
print(np.mean([score for score in lasso]))

0.8364857804832433


In [13]:
from sklearn.linear_model import LarsCV
lars = createModel(matrix, LarsCV(max_iter=100, cv=4, eps=5))
print(np.mean([score for score in lars]))

0.8395182374744704


In [14]:
from sklearn.linear_model import LassoLarsCV
lassoLars = createModel(matrix, LarsCV(max_iter=100, cv=4, eps=5))
print(np.mean([score for score in lassoLars]))

0.8395182374744704


#### Bonus question (15%)
Create a new user who represents your movie ratings. Rate some movies after your own taste and see how models 
evaluate non-selected movies. Do you find the predictions appropriate?

You can split the code needed for each answer into multiple cells.

### Notes

Implementation, description, and evaluation of supervised learning methods are included in libraries `skleaarn` or` Orange`.